src/db.c

```diff
@@ -2241,15 +2334,16 @@ int keyIsExpired(redisDb *db, robj *key) {
  *
  * The return value of the function is KEY_VALID if the key is still valid.
  * The function returns KEY_EXPIRED if the key is expired BUT not deleted,
- * or returns KEY_DELETED if the key is expired and deleted. */
-keyStatus expireIfNeeded(redisDb *db, robj *key, int flags) {
-    return expireIfNeededWithSlot(db,key,flags,getKeySlot(key->ptr));
-}
-
-static inline keyStatus expireIfNeededWithSlot(redisDb *db, robj *key, int flags, const int keySlot) {
+ * or returns KEY_DELETED if the key is expired and deleted.
+ * 
+ * You can optionally pass `kv` to save a lookup. 
+ */
+keyStatus expireIfNeeded(redisDb *db, robj *key, kvobj *kv, int flags) {
+    serverAssert(key != NULL);
+    sds keyname = key->ptr;
     if ((server.allow_access_expired) ||
         (flags & EXPIRE_ALLOW_ACCESS_EXPIRED) ||
-        (!keyIsExpiredWithSlot(db,key,keySlot)))
+        (!keyIsExpired(db, keyname, kv)))
```

Comment:
i saw that the changes from https://github.com/redis/redis/pull/13796 were revert by this PR, no longer needed? FYI @fcostaoliveira

Reply:
The expiration time is now embedded in kvobj, eliminating the need for a separate lookup. If `kv` is provided, we can return immediately. Else, a lookup is performed. This function is mainly called by `lookupKey()` and `delGenericCommand()`, where the former always supplies a non-NULL kv. Therefore, this change should improve performance.